ob2/util/job_limiter.py

```diff
@@ -0,0 +1,24 @@
+import logging
+
+from ob2.database import DbCursor
+from ob2.util.build_constants import QUEUED, IN_PROGRESS, FAILED
+from ob2.util.time import now_str
+
+MAX_JOBS_ALLOWED = 1
+
+def limit_queued_job(repo_name, build_name, limit):
+    try:
+        with DbCursor() as c:
+            c.execute('''SELECT count(*) FROM builds
+                         WHERE source = ? AND (status = ? OR status = ?)''',
+                      [repo_name, QUEUED, IN_PROGRESS])
+            count = int(c.fetchone()[0])
+            if count > limit:
```

Comment:
I think this ought to be in ob2.database.helpers.create_build. If there are too many queued, just raise an exception and don't create a new record in the builds table?

Reply:
I would rather have a build created and then failed, so the user can see / be redirected to a failed page with a reason as to why.

Reply:
That seems like a lazy way to accomplish the web UI + email notification for quota overages. Admission control should be separate from failures in the grading itself.

As an alternative, why not cancel old builds when new builds are scheduled? (with a specified concurrency limit) That's usually the intended behavior anyway.

Also, in the current implementation, wouldn't this silently cancel builds without sending any email notification? Is that intentional?

Reply:
Thats much more work than I have time to do - we just had an issue with groups spamming the autograder for to pass sync tests. I'm not actually sure that failing old builds is the right approach. Maybe we can schedule additional jobs by the same source in a separate low priority queue such that there's only ever MAX_JOBS_ALLOWED jobs in the normal queue + workers.

I'll pass these to TAs next semester or whoever I get to work on this.